contributors/design-proposals/storage/attacher-detacher-refactor-for-local-storage.md

```diff
@@ -0,0 +1,265 @@
+---
+
+title: Attacher/Detacher refactor for local storage
+
+authors:
+- "@NickrenREN"
+
+owning-sig: sig-storage
+
+participating-sigs:
+  - nil
+
+reviewers:
+  - "@msau42"
+  - "@jsafrane"
+
+approvers:
+  - "@jsafrane"
+  - "@msau42"
+  - "@saad-ali"
+
+editor: TBD
+
+creation-date: 2018-07-30
+
+last-updated: 2018-07-30
+
+status: provisional
+
+---
+
+## Table of Contents
+ * [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+* [Implementation](#implementation)
+    * [Volume plugin interface change](#volume-plugin-interface-change)
+    * [MountVolume/UnmountDevice generation function change](#MountVolume/UnmountDevice-generation-function-change)
+    * [Volume plugin change](#volume-plugin-change)
+* [Future](#future)
+
+## Summary
+
+Today, the workflow for a volume to be used by pod is:
+- attach a remote volume to the node instance (if it is attachable)
+- wait for the volume to be attached (if it is attachable)
+- mount the device to a global path (if it is attachable)
+- mount the global path to a pod directory
+
+It is ok for remote block storage plugins which have a remote attach api,such as `GCE PD`, `AWS EBS`
+and remote fs storage plugins such as `NFS`, and `Cephfs`.
+But it is not so good for plugins which need local attach such as `fc`, `iscsi` and `RBD`.
+It is not so good for local storage neither which is not attachable but needs `MountDevice`
+
+## Motivation
+
+### Goals
+
+ Update Attacher/Detacher interfaces for local storage
+
+### Non-Goals
+
+ Update `fc`, `iscsi` and `RBD` implementation according to the new interfaces
+
+## Proposal
+
+Here we propose to only update the Attacher/Detacher interfaces for local storage.
+We may expand it in future to `iscsi`, `RBD` and `fc`, if we figure out how to prevent multiple local attach without implementing attacher interface.
+
+## Implementation
+
+### Volume plugin interface change
+
+We can create a new interface `DeviceMounter`,  move `GetDeviceMountPath` and `GetDeviceMountPath` from `Attacher`to it.
+
+Any one who implement the `Attacher` interface must implement `DeviceMounter` interface.
+```
+// Attacher can attach a volume to a node.
+type Attacher interface {
+        // Attaches the volume specified by the given spec to the node with the given Name.
```

Comment:
Can you specify DeviceMounter interface inside the Attacher interface since we still require that anyone that implements Attach also implement MountDevice?

Reply:
done, thanks